### PR TITLE
8330677: Add Per-Compilation memory usage to JFR

### PR DESCRIPTION
**Pull request by tstuefe**
We have the (opt-in, disabled by default) compiler memory statistics introduced with [JDK-8317683](https://bugs.openjdk.org/browse/JDK-8317683).

Since temporary memory usage by compilers can significantly affect process footprint, it would make sense to expose at least the total peak usage per compilation via JFR.

--- 

This patch adds "Arena Usage" to CompilationEvent. We now see in JMC how costly a compilation had been. (The cost can get very high, as we have seen just recently again with [JDK-8327247](https://bugs.openjdk.org/browse/JDK-8327247) ).

jmc-memstat

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8330677](https://bugs.openjdk.org/browse/JDK-8330677): Add Per-Compilation memory usage to JFR (**Enhancement** - P4)

### Reviewers
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**) ⚠️ Review applies to [6829396e](https://git.openjdk.org/jdk/pull/18864/files/6829396effadd5420fa96bd9c53e81049425cccc)
 * [Matthias Baesken](https://openjdk.org/census#mbaesken) (@MBaesken - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/18864/head:pull/18864` \
`$ git checkout pull/18864`

Update a local copy of the PR: \
`$ git checkout pull/18864` \
`$ git pull https://git.openjdk.org/jdk.git pull/18864/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 18864`

View PR using the GUI difftool: \
`$ git pr show -t 18864`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/18864.diff">https://git.openjdk.org/jdk/pull/18864.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/18864#issuecomment-2067931325)

### CODE REVIEW

**Comment by tstuefe** [comment]
label hotspot-compiler

**Review by vnkozlov — APPROVED** [review]
Looks good.

**Comment by MBaesken** [comment]
Please also check COPYRIGHT years, e.g. compileTask.cpp .
Could you also check if there is an easy way to enhance some existing test for the JFR Compilation event. always good to have at least some basic testing .

**Comment by MBaesken** [comment]
> We have the (opt-in, disabled by default)

Might be helpful to add to the Description part of https://bugs.openjdk.org/browse/JDK-8317683 another sentence that describes how to enable the feature.

**Comment by tstuefe** [comment]
Thanks @MBaesken . Please check again to see if I addressed all your issues.

**Comment by MBaesken** [comment]
Regarding src/hotspot/share/jfr/metadata/metadata.xml , could you label it 'Arena Peak Usage' or something like this ? I would like to have it more clear that it is the peak usage.
Did you check my nullptr check related question ?
Otherwise looks okay to me.
It is a bit unfortunate that by default , we have now an 'empty' added field in the JFR Compilation event. But it is what it is; any chances that the compiler mem statistics would be enabled by default in the future, or is this considered to costly ?

**Comment by tstuefe** [comment]
> Regarding src/hotspot/share/jfr/metadata/metadata.xml , could you label it 'Arena Peak Usage' or something like this ? I would like to have it more clear that it is the peak usage. 

I already find "Arena Usage" to be too long, to be honest. Longer column labels make the JMC table less readable (ideally, one would have a "label," short and descriptive, and a "description," one being the column label, the other, e.g., a tooltip).

And "peak" is really the only option that makes sense here. If you ask someone what they think "usage" means, they will assume its the largest footprint accumulated during compilation over the time span of the compilation, aka peak.

> Did you check my nullptr check related question ? 

I read through the comments twice and did not find a nullptr related question. Which question?

> Otherwise looks okay to me. It is a bit unfortunate that by default , we have now an 'empty' added field in the JFR Compilation event. But it is what it is; any chances that the compiler mem statistics would be enabled by default in the future, or is this considered too costly ?

One synchronization per compilation. So, not that costly, no. It will be enabled by default in debug builds with https://github.com/openjdk/jdk/pull/18969 since it is implied in memlimit.

**Comment by MBaesken** [comment]
> I read through the comments twice and did not find a nullptr related question. Which question?

At some places we check the result for nullptr e.g.

jdk/src/hotspot/share/compiler/compilationMemoryStatistic.cpp

Line 79 in b3bcc49

	 const CompileTask* const task = th->task();

**Comment by tstuefe** [comment]
> > I read through the comments twice and did not find a nullptr related question. Which question?
> 
> See compilationMemoryStatistic.cpp .
> 
> At some places we check the result for nullptr e.g.
> 
> jdk/src/hotspot/share/compiler/compilationMemoryStatistic.cpp
> 
> Line 79 in [b3bcc49](https://github.com/openjdk/jdk/commit/b3bcc49491b8f8ad337eb4c06201a5468e5c1159)
> 
> ```
>  const CompileTask* const task = th->task(); 
> ```

Yes, it is inconsistent. Allmost all code here (notably anything triggered from start- or end-compilation) are called from the compiler so we run on a compiler thread and in the scope of a ciEnv. So most of the existing nullptr checks are probably not needed. 

We may want to make this consistent with subsequent RFEs.

**Comment by MBaesken** [comment]
> We may want to make this consistent with subsequent RFEs.

I agree, this can be a follow up.

**Review by MBaesken — APPROVED** [review]

File: src/hotspot/share/jfr/metadata/metadata.xml [hunk 1]
```diff
@@ -608,6 +608,7 @@
     <Field type="boolean" name="isOsr" label="On Stack Replacement" />
     <Field type="ulong" contentType="bytes" name="codeSize" label="Compiled Code Size" />
     <Field type="ulong" contentType="bytes" name="inlinedBytes" label="Inlined Code Size" />
+    <Field type="ulong" contentType="bytes" name="arenaBytes" label="Arena Usage" />
```
[the remark this answers is not in the document]
Comment (tstuefe) [reply src/hotspot/share/jfr/metadata/metadata.xml:611]: Again, there is no space for this. "label" gets used as column header. No space to add a descriptive text.

**Comment by tstuefe** [comment]
Thanks @MBaesken . For some reason your remarks only got posted now, but I think we covered all points already.

/integrate

**Pull request closed without merging**